pyscf/solvent/grad/pcm.py

```diff
@@ -139,173 +138,256 @@ def get_dD_dS(surface, dF, with_S=True, with_D=False):
+    dvj= 2.0 * numpy.asarray([numpy.sum(dvj[:,p0:p1], axis=1) for p0,p1 in aoslice[:,2:]])
+    de = dq + dvj
+    t1 = log.timer_debug1('grad qv', *t1)
+    return de
+
+
+def grad_solver(pcmobj, dm):
+    '''
+    dE = 0.5*v* d(K^-1 R) *v + q*dv
+    v^T* d(K^-1 R)v = v^T*K^-1(dR - dK K^-1R)v = v^T K^-1(dR - dK q)
+    '''
+    mol = pcmobj.mol
+    log = logger.new_logger(mol, mol.verbose)
+    t1 = (logger.process_clock(), logger.perf_counter())
+    if not pcmobj._intermediates:
+        pcmobj.build()
+    dm_cache = pcmobj._intermediates.get('dm', None)
+    if dm_cache is not None and numpy.linalg.norm(dm_cache - dm) < 1e-10:
+        pass
+    else:
+        pcmobj._get_vind(dm)
+
+    gridslice    = pcmobj.surface['gslice_by_atom']
+    v_grids      = pcmobj._intermediates['v_grids']
+    A            = pcmobj._intermediates['A']
+    D            = pcmobj._intermediates['D']
+    S            = pcmobj._intermediates['S']
+    K            = pcmobj._intermediates['K']
+    q            = pcmobj._intermediates['q']
+
+    vK_1 = numpy.linalg.solve(K.T, v_grids)
 
-    ## --------------- response from stiffness matrices ----------------
-    gridslice = pcmobj.surface['gslice_by_atom']
     dF, dA = get_dF_dA(pcmobj.surface)
 
-    with_D = pcmobj.method.upper() == 'IEF-PCM' or pcmobj.method.upper() == 'SS(V)PE'
+    with_D = pcmobj.method.upper() in ['IEF-PCM', 'IEFPCM', 'SS(V)PE', 'SMD']
     dD, dS, dSii = get_dD_dS(pcmobj.surface, dF, with_D=with_D, with_S=True)
 
-    if pcmobj.method.upper() == 'IEF-PCM' or pcmobj.method.upper() == 'SS(V)PE':
+    if pcmobj.method.upper() in ['IEF-PCM', 'IEFPCM', 'SS(V)PE', 'SMD']:
         DA = D*A
 
     epsilon = pcmobj.eps
 
     #de_dF = v0 * -dSii_dF * q
     #de += 0.5*numpy.einsum('i,inx->nx', de_dF, dF)
     # dQ = v^T K^-1 (dR - dK K^-1 R) v
-    if pcmobj.method.upper() == 'C-PCM' or pcmobj.method.upper() == 'COSMO':
+    de = numpy.zeros([pcmobj.mol.natm,3])
+    if pcmobj.method.upper() in ['C-PCM', 'CPCM', 'COSMO']:
+        dS = dS.transpose([2,0,1])
+        dSii = dSii.transpose([2,0,1])
+
         # dR = 0, dK = dS
-        de_dS = numpy.einsum('i,ijx,j->ix', vK_1, dS, q)
+        de_dS = (vK_1 * dS.dot(q)).T                  # cupy.einsum('i,xij,j->ix', vK_1, dS, q)
         de -= numpy.asarray([numpy.sum(de_dS[p0:p1], axis=0) for p0,p1, in gridslice])
-        de -= 0.5*numpy.einsum('i,ijx,i->jx', vK_1, dSii, q)
+        de -= 0.5*numpy.einsum('i,xij->jx', vK_1*q, dSii) # 0.5*cupy.einsum('i,xij,i->jx', vK_1, dSii, q)
+
+    elif pcmobj.method.upper() in ['IEF-PCM', 'IEFPCM', 'SS(V)PE', 'SMD']:
+        dD = dD.transpose([2,0,1])
+        dS = dS.transpose([2,0,1])
+        dSii = dSii.transpose([2,0,1])
+        dA = dA.transpose([2,0,1])
+        def contract_bra(a, B, c):
+            ''' i,xij,j->jx '''
+            tmp = numpy.einsum('i,xij->xj', a, B)
+            return (tmp * c).T
+
+        def contract_ket(a, B, c):
```

Comment:
Using the einsum version for `contract_bra` and `contract_ket` seems better. Is there big performance difference?